keps/sig-network/4858-ip-cidr-validation/README.md

```diff
@@ -0,0 +1,838 @@
+KEP is to make that impossible.
+
+(I guess possibly we might be able to set up the e2e test to be able
+to write the object directly into etcd, bypassing the apiserver? Or
+maybe add an "e2e test run" mode to the apiserver, that makes it
+possible to request that it skip validation for a particular
+Create/Update? Alternatively, this is an argument for eventually
+forcibly fixing all remaining invalid values in etcd, since it will no
+longer be possible to test behavior regarding them.)
+
+[equivalent unit tests]: https://github.com/kubernetes/kubernetes/pull/126203
+
+### Graduation Criteria
+
+#### Alpha
+
+- API warnings added.
+
+- Mandatory strict validation for `NetworkDeviceData`.
+
+- All kube-internal controllers only write out canonical IPs/CIDRs.
+
+- Updated validation available behind a feature gate, but disabled by
+  default.
+
+- Unit tests added/updated.
+
+#### Beta
+
+- Updated validation available behind a feature gate, enabled by
+  default.
+
+- Unit tests added/updated, old e2e and integration tests removed.
+
+#### GA
+
+- Allow time for feedback, etc
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+N/A; No interactions between kubernetes components should be affected
+by the change, so skew doesn't matter.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: StrictIPCIDRValidation
+  - Components depending on the feature gate:
+    - kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+
+Yes, it changes the validation of multiple fields.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Nothing surprising.
+
+###### Are there any tests for feature enablement/disablement?
+
+No
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+Rollback should not be able to fail.
+
+Rollout would fail if the cluster contains objects with invalid
+IPs/CIDRs that must be re-created as part of the rollout/upgrade. For
+example, if kube-proxy was deployed via a DaemonSet whose template
+contained an invalid IP in `.spec.dnsConfig.nameservers[]`, then it
+would not be possible to recreate the pod after the apiserver was
+updated.
+
+The warnings in the Alpha period will hopefully avert this problem.
+
+###### What specific metrics should inform a rollback?
+
+None; if anything fails, it is likely to involve user workloads. There
+are not any metrics for validation failures (but see the "missing
+metrics" question below).
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+No
```

Comment:
Not required for alpha, but for beta/GA you'll need to preform manual upgrade->downgrade->upgrade testing and describe the steps here.